rpcs3/Emu/RSX/rsx_cache.h

```diff
@@ -695,24 +714,26 @@ namespace rsx
 			std::string filename = fmt::format("%llX.fp", program_hash);
 
 			fs::file f(root_path + "/raw/" + filename);
-			f.read<u8>(data, f.size());
+			if (f) f.read(data, f.size());
 
 			RSXFragmentProgram fp = {};
+			fp.ucode_length = ::size32(data);
 			{
-				std::lock_guard<std::mutex> lock(fpd_mutex);
-				fragment_program_data[program_hash] = data;
-				fp.data = fragment_program_data[program_hash].data();
+				std::lock_guard lock(fpd_mutex);
+				fp.data = fragment_program_data.insert_or_assign(program_hash, std::move(data)).first->second.data();
 			}
-			fp.ucode_length = ::size32(data);
 
 			return fp;
 		}
 
 		std::tuple<pipeline_storage_type, RSXVertexProgram, RSXFragmentProgram> unpack(pipeline_data &data)
 		{
-			RSXVertexProgram vp = load_vp_raw(data.vertex_program_hash);
-			RSXFragmentProgram fp = load_fp_raw(data.fragment_program_hash);
-			pipeline_storage_type pipeline = data.pipeline_properties;
+			std::tuple<pipeline_storage_type, RSXVertexProgram, RSXFragmentProgram> result;
```

Comment:
What does this rewrite try to achieve?

Reply:
copy elision.

Reply:
The compiler couldn't optimize it correctly? It seems to be very basic to me.

Reply:
Or maybe I should've forwarded tuple with move instead of make.